nav2_voxel_grid/src/voxel_grid.cpp

```diff
@@ -0,0 +1,255 @@
+    return;
+  }
+
+  MarkVoxel mv(data_);
+  raytraceLine(mv, x0, y0, z0, x1, y1, z1, max_length);
+}
+
+void VoxelGrid::clearVoxelLine(
+  double x0, double y0, double z0, double x1, double y1, double z1,
+  unsigned int max_length)
+{
+  if (x0 >= size_x_ || y0 >= size_y_ || z0 >= size_z_ || x1 >= size_x_ || y1 >= size_y_ ||
+    z1 >= size_z_)
+  {
+    RCLCPP_DEBUG(logger,
+      "Error, line endpoint out of bounds. "
+      "(%.2f, %.2f, %.2f) to (%.2f, %.2f, %.2f),  size: (%d, %d, %d)",
+      x0, y0, z0, x1, y1, z1, size_x_, size_y_, size_z_);
+    return;
+  }
+
+  ClearVoxel cv(data_);
+  raytraceLine(cv, x0, y0, z0, x1, y1, z1, max_length);
+}
+
+void VoxelGrid::clearVoxelLineInMap(
+  double x0, double y0, double z0, double x1, double y1, double z1, unsigned char * map_2d,
+  unsigned int unknown_threshold, unsigned int mark_threshold, unsigned char free_cost,
+  unsigned char unknown_cost, unsigned int max_length)
+{
+  costmap = map_2d;
+  if (map_2d == NULL) {
+    clearVoxelLine(x0, y0, z0, x1, y1, z1, max_length);
+    return;
+  }
+
+  if (x0 >= size_x_ || y0 >= size_y_ || z0 >= size_z_ || x1 >= size_x_ || y1 >= size_y_ ||
+    z1 >= size_z_)
+  {
+    RCLCPP_DEBUG(logger,
+      "Error, line endpoint out of bounds. "
+      "(%.2f, %.2f, %.2f) to (%.2f, %.2f, %.2f),  size: (%d, %d, %d)",
+      x0, y0, z0, x1, y1, z1, size_x_, size_y_, size_z_);
+    return;
+  }
+
+  ClearVoxelInMap cvm(data_, costmap, unknown_threshold, mark_threshold, free_cost, unknown_cost);
+  raytraceLine(cvm, x0, y0, z0, x1, y1, z1, max_length);
+}
+
+VoxelStatus VoxelGrid::getVoxel(unsigned int x, unsigned int y, unsigned int z)
+{
+  if (x >= size_x_ || y >= size_y_ || z >= size_z_) {
+    RCLCPP_DEBUG(logger, "Error, voxel out of bounds. (%d, %d, %d)\n", x, y, z);
+    return UNKNOWN;
+  }
+  uint32_t full_mask = ((uint32_t)1 << z << 16) | (1 << z);
+  uint32_t result = data_[y * size_x_ + x] & full_mask;
+  unsigned int bits = numBits(result);
+
+  // known marked: 11 = 2 bits, unknown: 01 = 1 bit, known free: 00 = 0 bits
+  if (bits < 2) {
+    if (bits < 1) {
+      return FREE;
+    }
+
+    return UNKNOWN;
+  }
+
+  return MARKED;
+}
+
+VoxelStatus VoxelGrid::getVoxelColumn(
+  unsigned int x, unsigned int y,
+  unsigned int unknown_threshold, unsigned int marked_threshold)
+{
+  if (x >= size_x_ || y >= size_y_) {
+    RCLCPP_DEBUG(logger, "Error, voxel out of bounds. (%d, %d)\n", x, y);
+    return UNKNOWN;
+  }
+
+  uint32_t * col = &data_[y * size_x_ + x];
+
+  unsigned int unknown_bits = uint16_t(*col >> 16) ^ uint16_t(*col);
+  unsigned int marked_bits = *col >> 16;
+
+  // check if the number of marked bits qualifies the col as marked
+  if (!bitsBelowThreshold(marked_bits, marked_threshold)) {
+    return MARKED;
+  }
+
+  // check if the number of unkown bits qualifies the col as unknown
+  if (!bitsBelowThreshold(unknown_bits, unknown_threshold)) {
+    return UNKNOWN;
+  }
+
+  return FREE;
+}
+
+unsigned int VoxelGrid::sizeX()
```

Comment:
Seems inconsistent to me to have these little functions in the .cpp file while there are much larger methods in the header file. I would tend to put the little ones in the header (inline), while the larger would be in the .cpp file.

Reply:
I expect there is plenty of room for improvement in this code. But this was just a simple port of the voxel grid package. The code is unchanged from ROS 1 except to compile under ROS 2 and pass tests. Making it better wasn't part of this PR